### ISSUE TITLE
[Dynamo] bytecode transformed by Dynamo is not serializable by marshal

### ISSUE THREAD

**Issue by youkaichao** [pytorch/pytorch#116013]
### 🐛 Describe the bug

By default, Python code object uses [`marshal`](https://docs.python.org/3/library/marshal.html) for serialization. However, I find that bytecode transformed by Dynamo sometimes is not serializable by `marshal`.

The reason is that Dynamo stores a class variable inside `co_consts`.

Finally, I managed to use [`dill`](https://pypi.org/project/dill/) to save the code object. For those who wants to see the live object, please download it [here](https://cloud.tsinghua.edu.cn/f/7f9dc5f9c27f4076bd56/?dl=1), and use the code below to load it:

```python
import dill
code = dill.load(open("__transformed_code_0_for_resume_in_forward.py.transformed_bytecode", "rb"))
```

The object in question is `code.co_consts[7]`, which is a `transformers.models.longformer.modeling_longformer.LongformerMaskedLMOutput` class.

The class variable also causes problems for debuggers. When I view the code object in debuggers (I'm using VS Code), these variables are invisible. The following screenshot shows the problem: the object in index `7` is not shown.

<img width="760" alt="image" src="https://github.com/pytorch/pytorch/assets/23236638/1fa32425-dff5-469f-b6c1-5e5b3429fd11">

In summary, while storing class variables in `co_consts` works for Python VM execution, it might not be compatible with the Python ecosystem.

### Versions

I'm using PyTorch version `2.2.0.dev20231213`.

**Comment by youkaichao** [comment]
Here is a minimal working example:

```python
def f():
    return 1
class A(object):
    pass
new_code = f.__code__.replace(co_consts=(2, A))
f.__code__ = new_code
print(f())
```

The above code can run without any problem, but if we try to `marshal` it, it will raise an exception.

**Comment by youkaichao** [comment]
cc @jansel

**Comment by jansel** [comment]
Seems like we should change the way we store that class, though I'm not sure that there are supported usecases for serializing generated bytecode.

**Comment by youkaichao** [comment]
My usecase is to dump bytecode so that I can analyze the bytecode and run decompile-compile "offline" in CPU only machines (my laptop).

**Comment by youkaichao** [comment]
I think the relevant code is:

https://github.com/pytorch/pytorch/blob/2a440348958b3f0a2b09458bd76fe5959b371c0c/torch/_dynamo/utils.py#L894

Dynamo thinks these types are safe as constant:

```python
def is_safe_constant(v):
    if istype(v, (tuple, frozenset)):
        return all(map(is_safe_constant, v))
    return isinstance(v, (enum.Enum, type)) or istype(
        v,
        (
            types.CodeType,
            int,
            float,
            bool,
            str,
            bytes,
            type(None),
            slice,
            type(type),
            torch.device,
            torch.dtype,
        ),
    )
```

And there are some weird code that pushes a function object in `co_consts`:

https://github.com/pytorch/pytorch/blob/2a440348958b3f0a2b09458bd76fe5959b371c0c/torch/_dynamo/codegen.py#L297

If we want to ensure the bytecode can be serializable by `marshal`, that would break lots of code and need quite a big code change. We need to place these objects in global namespace, and issue a `LOAD_GLOBAL` instruction instead.

If anyone wants to serialize the bytecode, just using `dill` does the trick.

In the long run, I think using `LOAD_GLOBAL` would be better. `co_consts` is the place for constant variables that can be represented safely by a piece of source code.

**Comment by voznesenskym** [comment]
What are you doing that you need this?

**Comment by voznesenskym** [comment]
> My usecase is to dump bytecode so that I can analyze the bytecode and run decompile-compile "offline" in CPU only machines (my laptop).

Why do you need serialization to this? Would logging out the bytecode as it gets generated suffice?

**Comment by youkaichao** [comment]
The logging such as `dis.dis(code)` is lossy, and it is not easy to collect `co_consts/co_nlocals/co_varnames` etc. And the most important thing: decompilation requires a code object.
